crates/c-api/include/wasmtime/async.h

```diff
@@ -0,0 +1,254 @@
+ * \brief Whether or not to enable support for asynchronous functions in Wasmtime.
+ *
+ * When enabled, the config can optionally define host functions with async. 
+ * Instances created and functions called with this Config must be called through their asynchronous APIs, however.
+ * For example using wasmtime_func_call will panic when used with this config.
+ *
+ * For more information see the Rust documentation at
+ * https://docs.wasmtime.dev/api/wasmtime/struct.Config.html#method.async_support
+ */
+WASMTIME_CONFIG_PROP(void, async_support, bool)
+
+/**
+ * \brief Configures the size of the stacks used for asynchronous execution.
+ *
+ * This setting configures the size of the stacks that are allocated for asynchronous execution. 
+ *
+ * The value cannot be less than max_wasm_stack.
+ *
+ * The amount of stack space guaranteed for host functions is async_stack_size - max_wasm_stack, so take care
+ * not to set these two values close to one another; doing so may cause host functions to overflow the stack 
+ * and abort the process.
+ *
+ * By default this option is 2 MiB.
+ *
+ * For more information see the Rust documentation at
+ * https://docs.wasmtime.dev/api/wasmtime/struct.Config.html#method.async_stack_size
+ */
+WASMTIME_CONFIG_PROP(void, async_stack_size, uint64_t)
+
+/**
+ * \brief Configures a Store to yield execution of async WebAssembly code periodically.
+ *
+ * When a Store is configured to consume fuel with #wasmtime_config_consume_fuel 
+ * this method will configure what happens when fuel runs out. Specifically executing
+ * WebAssembly will be suspended and control will be yielded back to the caller.
+ *
+ * This is only suitable with use of a store associated with an async config because
+ * only then are futures used and yields are possible.
+ */
+WASM_API_EXTERN void wasmtime_context_out_of_fuel_async_yield(
+    wasmtime_context_t *context,
+    uint64_t injection_count,
+    uint64_t fuel_to_inject);
+
+/**
+ * \brief Configures epoch-deadline expiration to yield to the async caller and the update the deadline.
+ *
+ * This is only suitable with use of a store associated with an async config because
+ * only then are futures used and yields are possible.
+ *
+ * See the Rust documentation for more:
+ * https://docs.wasmtime.dev/api/wasmtime/struct.Store.html#method.epoch_deadline_async_yield_and_update
+ */
+WASM_API_EXTERN void wasmtime_context_epoch_deadline_async_yield_and_update(wasmtime_context_t *context, uint64_t delta);
+
+/**
+ * The callback to determine a continuation's current state.
+ *
+ * Return true if the host call has completed, otherwise false will 
+ * continue to yield WebAssembly execution.
+ *
+ * \param if error is assigned a non `NULL` value then the called function will 
+ *        trap with the returned error. Note that ownership of error is transferred 
+ *        to wasmtime.
+ */
+typedef bool (*wasmtime_func_async_continuation_callback_t)(
+    void *env,
+    wasm_trap_t **error);
```

Comment:
Does it make more sense for this error_ret to be a parameter of the initial host function call? Then this becomes the mirror to poll which just returns a boolean.

Reply:
I went ahead and did this - I quite like how it turned out.